### PR TITLE
Vkesav/dev

### PR DESCRIPTION
**Pull request by kesav-v**
Initial implementation of BERT layer for generating embeddings based on text data.

### CODE REVIEW

File: bert_trainer.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        self.tokenizer = _create_tokenizer_from_hub_module()
+
+        # Convert our train and test features to InputFeatures that BERT understands.
+        if self.is_training:
+            self.train_features = bert.run_classifier.convert_examples_to_features(
+                train_input_examples, self.label_list, self.max_seq_length, self.tokenizer)
+            self.test_features = bert.run_classifier.convert_examples_to_features(
+                test_input_examples, self.label_list, self.max_seq_length, self.tokenizer)
+
+    def _create_model(self, is_predicting, input_ids, input_mask, segment_ids, labels,
```
Comment (siqiz-goog) [inline bert_trainer.py:93]: Please add some comments on the input arguments in the style defined by
http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: get_nyt_articles.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+'''
+Reads in articles from the New York Times API and saves
+them to a cache.
+'''
+
+import time
```
Comment (siqiz-goog) [inline get_nyt_articles.py:6]: Could you run pylint on this part, I think it can auto-format the the imports better 

http://google.github.io/styleguide/pyguide.html#21-lint

File: bert_trainer.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    def test_model(self):
+        test_input_fn = run_classifier.input_fn_builder(
+                features=self.test_features,
+                seq_length=self.max_seq_length,
+                is_training=False,
+                drop_remainder=False)
+
+        return self.estimator.evaluate(input_fn=test_input_fn, steps=None)
+
+    def predict(self, inputs):
```
Comment (siqiz-goog) [inline bert_trainer.py:249]: add some comments about the inputs

http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: bert_trainer.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            per_example_loss = -tf.reduce_sum(one_hot_labels * log_probs, axis=-1)
+            loss = tf.reduce_mean(per_example_loss)
+            return (loss, predicted_labels, log_probs)
+
+    # model_fn_builder actually creates our model function
+    # using the passed parameters for num_labels, learning_rate, etc.
+    def _model_fn_builder(self, num_labels, learning_rate, num_train_steps,
+            num_warmup_steps):
+        """Returns `model_fn` closure for TPUEstimator."""
+        def model_fn(features, labels, mode, params):  # pylint: disable=unused-argument
```
Comment (siqiz-goog) [inline bert_trainer.py:152]: Add some comments about features and params
http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: bert_trainer.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+        print(output_layer.shape, output_weights.shape)
+
+        output_bias = tf.get_variable(
+                "output_bias", [num_labels], initializer=tf.zeros_initializer())
+
+        with tf.variable_scope("loss"):
+
+            # Dropout helps prevent overfitting
+            output_layer = tf.nn.dropout(output_layer, keep_prob=0.9)
```
Comment (siqiz-goog) [inline bert_trainer.py:128]: Should this be a hyper parameter?

Comment (siqiz-goog) [inline bert_trainer.py:128]: Also I believe usually 0.5 is used for dropout, is there a reason why we use 0.9

File: bert_trainer.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
```
Comment (breitenm-google) [inline bert_trainer.py:129]: The dropout should be skipped when predicting, right?

File: bert_trainer_example.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        save_checkpoints_every=args.save_checkpoints_every,
+        save_summary_every=args.save_summary_every,
+        dropout_rate=args.dropout_rate,
+        output_dir=args.output_dir,
+        is_training=args.training,
+    )
+    trainer.train_model()
+    eval_accuracy_info = trainer.test_model()
+    print(eval_accuracy_info)
+    preds = trainer.predict([
```
Comment (siqiz-goog) [inline bert_trainer_example.py:53]: Might want to leave out those examples into a test or something

File: get_nyt_articles.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            continue
+        if 'snippet' not in i.keys():
+            continue
+        dic = {}
+        dic['id'] = i['_id']
+        if i.get('abstract', 'EMPTY') is not None:
+            dic['abstract'] = i.get('abstract', 'EMPTY').encode("utf8")
+        dic['headline'] = i['headline']['main'].encode("utf8")
+        dic['desk'] = i.get('news_desk', 'EMPTY')
+        dic['date'] = i['pub_date'][0:10] # cutting time of day.
```
Comment (siqiz-goog) [inline get_nyt_articles.py:41]: Do some length checking here

File: get_nyt_articles.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+def bulk_look_up(start_year):
+    # create a list of year, month, pairs for the data
+    # from start dt to end date inclusive
+    start_dt = datetime.date(start_year, 1, 1)
+    end_dt = datetime.datetime.today()
+    dates = [(dt.year, dt.month) for dt in rrule(MONTHLY, dtstart=start_dt, until=end_dt)]
+
+    dfs = []
+    for year, month in dates:
+        url = (
```
Comment (siqiz-goog) [inline get_nyt_articles.py:67]: Add a comment to link to the source data api link

File: get_nyt_articles.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+def download_or_cache(start_year):
+    cache_path = 'nyt_data_from_{start_year}.pkl'.format(start_year=start_year)
+    if os.path.exists(cache_path):
+        return pd.read_pickle(cache_path)
+    df = bulk_look_up(start_year)
+    df.to_pickle(cache_path)
+    return df
+
+if __name__ == '__main__':
+    download_or_cache(2015)
```
Comment (siqiz-goog) [inline get_nyt_articles.py:88]: maybe make this a flag

File: bert_trainer_example.py (hunk 7 above)
Comment (kesav-v) [reply bert_trainer_example.py:55]: So in my tests i use a very small amount of toy data just to verify functionality. I thought this could be separate since it runs the full pipeline on all the NYT articles.

File: get_nyt_articles.py (hunk 8 above)
Comment (kesav-v) [reply get_nyt_articles.py:44]: I've added a check for the length of the "pub_date" field in the response.
